docs: bytecode storage (#372)
kkrt-labs/kakarot-ssj: docs/general/contract_bytecode.md

# Bytecode Storage Methods for Kakarot on Starknet

The bytecode is the compiled version of your contract, and it is what the

> Kakarot EVM will execute when you call the contract. As Kakarot is developed on

Reviewer: We would also benefit from being more precise, such as that Kakarot's state is embedded into Starknet's state

Reviewer: That way we're not tied to saying we're deployed on Starknet L2 (and saying we're developed on top of Starknet implies that imo

> top of Starknet, you cannot really "deploy" an EVM contract on Kakarot: what

Reviewer: i think we would benefit from rephrasing this
in a way, 
- running a constructor bytecode
- storing on-chain the deployed bytecode of a contract and making it accessible at a particular address 

is the essence of deployment

what we're not enabling are deploy txs in some sense?

actually happens is that the EVM bytecode of your contract is stored on the
blockchain, and the Kakarot EVM will be able to load it when you want to execute

> it.

Reviewer: Lgtm, I think overall we can rephrase the content to not contain "you" and "your" but speak passively?
Maybe claude can help, or gpt

Author: doing this after first wave of changes

  contract's storage.

These three solutions all have their pros and cons, and we will go over them in
the following sections.

## Foreword: Data availability

In Validity Rollups, verifying the validity proof on L1 is sufficient to
guarantee the validity of a transaction execution on L2, with no need to have
the detailed transaction information sent to Ethereum.

However, in order to allow the independent verification of the L2 chain's state
and prevent malicious operators from censoring or freezing the chain, some
amount of data is still required to be posted on a Data Availability (DA) layer
to make the Starknet state available, even in the case where the operator
suddenly ceases operations. Data availability ensures that users can always
reconstruct the state of the rollup by deriving its current state from the data
posted by the rollup operator.

Without this, users would not be able to query an L2 contract's state in case
the operator becomes unavailable. It provides users the security of knowing that
if the Starknet sequencer ever stops functioning, they can prove custody of
their funds using the data posted on the DA Layer. If that DA Layer is Ethereum

> itself, then they inherit from Ethereum's security guarantees.

Reviewer: nice

## Different approaches to storing contract bytecode

### Using Ethereum as a DA Layer

Starknet currently uses Ethereum as its DA Layer. Each state update verified
on-chain is accompanied by the state diff between the previous and new state,
sent as calldata to Ethereum, allowing anyone that observes Ethereum to
reconstruct the current state of Starknet. This security comes with a
significant price, as the publication of state diffs on Ethereum accounted for
[over 93% of the transaction fees paid on Starknet](https://community.starknet.io/t/volition-hybrid-data-availability-solution/97387).

The first choice when it comes to storing contract bytecode is to store it as a
regular storage variable, whose state diff is posted on Ethereum acting as the
DA Layer. As outlined in our [Contract Storage](./contract_storage.md) design,
deploying a new contract on Kakarot would not result in the deployment of a
contract on Starknet, but rather in the storage of the contract bytecode in a

> storage variable of the KakarotCore contract.

Reviewer: very clear, nice

actual state of the L2 is. By posting state diffs on L1, we can reconstruct the
current state of Starknet from the ground up, but this has a significant cost.

![Volition](volition.png)

Volition will allow developers to choose whether data will be stored in the L1DA
or L2DA mode, making it possible to store data on L2, which is a lot less
expensive than storing it on L1. Depending on the data stored, it can be
interesting if the cost associated to storing the data on L1 is higher than the
intrinsic value of the data itself. For examples, an Volition-ERC20 token
standard would have two different balances stored, one on L1DA for maximal
security (e.g. you would keep most of your assets in this balance), and one on
L2DA for lower security, which would be used to reduce the fees associated to
small transactions.

In our case, we would store the contract bytecode in a storage variable that is
settled on the L2DA instead of the L1DA. This would make contract deployment
extremely cheap on Kakarot, as we will save the cost of posting the state diff
associated to the update of our stored bytecode on Ethereum.

#### Associated Risks

> There are some risks that must be considered when using Volition. Consider the

Reviewer: could someone from starkware volition team read this paragraph and give their opinion?
i still dont understand why when Starknet gets a consensus, Sequencers can sign that they had the required L2DA data to perform the state change 

which means every single sequencer in the consensus would have to go rogue and refuse to share the updated L2DA 

ofc its different for the centralized sequencer case

Author: > Starknet gets a consensus, Sequencers can sign that they had the required L2DA data to perform the state change

which means every single sequencer in the consensus would have to go rogue and refuse to share the updated L2DA

Not every: a malicious majority is sufficient.

Reviewer: > > Starknet gets a consensus, Sequencers can sign that they had the required L2DA data to perform the state change
> 
> which means every single sequencer in the consensus would have to go rogue and refuse to share the updated L2DA
> 
> Not every: a malicious majority is sufficient.

Why?

If just one sequencer shares its L2DA with every full node, the state of the chain is "safe"?

Author: > one sequencer shares its L2DA with every full node, the state of the chain is "safe"?

The sequencer is in minority. After the hostile takeover, the chain has agreed on a new L2DA root. If an information is hidden from honest sequencers, they can't compute the new state as they're missing information.

Reviewer: But this new L2DA root cannot be proven, correct? Because it cannot result in a valid proof

Author: The thing is that it actually can be proven (given that it's valid). You know that the state transition from S_A to S_B is valid. But if you don't make S_B available to validators, then how can they validate that the S_B -> S_C transition is valid? They can't, hence txs executing L2DA updates will fail. (As I understood it)

case of an attack by a majority of malicious sequencers colluding who decide to
not share a change in the L2DA with other sequencers and full nodes. Once the
attack is finished, the honest sequencers won't have the data needed to
reconstruct and compute the new root of the L2DA. In a such situation, not only
the L2DA is not accessible anymore, but any execution relying on L2DA will not
be executable and provable anymore, as sequencers won't have access the the L2DA
state.

Even though this event is unlikely to happen, it remains a possibility that must
be taken into account as L2DA is less secure than L1DA. If an event like this
were ever to happen, then the stored bytecode would be lost, and the deployed

> contract would not be executable anymore.

Reviewer: people could give the bytecode themself, as they are likely to "have" it locally forever

Author: The problem is not that the bytcode can't be accessed - it's that no operation using L2DA will succeed due to a failure of proving the current L2DA state.

The only migtigation would be to completely erase the L2DA state.

> Note: While we could potentially use Volition to store the bytecode on L2DA in
> the future, this is not possible at the moment, as Volition is not yet
> implemented on Starknet.

### Storing the EVM bytecode in the Cairo contract code

The last option is to store the EVM bytecode directly in the Cairo contract
code. This has the advantage of also being cheap, as this data is not stored on
L1.

On Starknet, there is a distinction between classes which is the definition of a
contract containing the Cairo bytecode, and contracts which are instances of
classes. When you declare a contract on Starknet, its information is added to
the
[Classes Tree](https://docs.starknet.io/documentation/architecture_and_concepts/Network_Architecture/starknet-state/#classes_tree),
which encodes information about the existing classes in the state of Starknet by

> mapping class hashes to their. This class tree is itself a part of the Starknet

Reviewer: mapping class hashes to their?
to their class?

Author: compiled class hash

State Commitment, which is verified on Ethereum during state updates. The class
itself is stored in the nodes (both sequencers and full nodes) of Starknet.

To implement this solution, we must declare a new class each time a contract is
deployed via Kakarot. This new class would contain the EVM bytecode of the
contract, exposed inside a view function that would return the entire bytecode
when queried. To achieve that, we would need to have the RPC craft a custom
Starknet contract that would contain this EVM bytecode, and declare it on

> Starknet - which is not ideal from security perspectives.

Reviewer: right after this paragraph, it's not so clear that we're going with option 1

except the first sentence of the next paragraph

Author: Added this information at the end of Solution 1

## Implementation details

Kakarot uses the first solution, storing the bytecode inside a storage variable
that is committed to Ethereum. This solution is the most secure one, as it
relies on Ethereum as a DA Layer, and thus inherits from Ethereum's security
guarantees, ensuring that the bytecode of the deployed contract is always
available.

> A `deploy` transaction is identified by the `to` address being set to zero. The

Reviewer: is this true? maybe it's if the address is not even set, or like `to: ""`
i'd have to check again 
but i think sending a tx to address(0) is valid

Author: fixed to: `being empty`

> data sent to the KakarotCore contract when deploying a new contract wil be

Reviewer: ```suggestion
data sent to the KakarotCore contract when deploying a new contract will be
```

formatted by the RPC to pack the bytecode into 31-bytes values, and passed as a

> `Array<felt252>` to the entrypoint `eth_call` of the KakarotCore contract. This

Reviewer: ```suggestion
`Array<felt252>` to the entrypoint `eth_send_transaction` of the KakarotCore contract. This
```

Reviewer: eth_call can only be called for view purposes atm

allows us to save on computation costs required to pack all byte values into the
31-bytes values that we will store in the contract storage.

The contract storage related to the deployed contract will be organized as such:

```rust
struct Storage {

>     bytecode: LegacyMap<EthAddress, List<bytes31>>,

Reviewer: do we also need valid jumps destinations here?

Author: I think so. I'm not sure how yet. Will be for another PR imo

    pending_word: LegacyMap<EthAddress, felt252>,
    pending_word_len: LegacyMap<EthAddress, usize>,
}
```

Each deployed contract has it's own EVM address, that we use as a key to a
`LegacyMap` type when computing the address of each storage variable. We use the

> `List` type from Alexandria to store the bytecode, as it allows us to store up

Reviewer: ```suggestion
`List` type from [Alexandria](https://github.com/keep-starknet-strange/alexandria/blob/main/src/storage/src/list.cairo) to store the bytecode, as it allows us to store up
```

to 255 31-bytes values, per `StorageBaseAddress`. For bytecode containing more
than 255 31-bytes values, the `List` type abstracts the calculations of the next
storage address used, which is calculated by using poseidon hashes applied on
`previous_address+1`.

The logic behind this storage design is to make it very easy to load the
bytecode in the EVM when we want to execute a program. We will rely on the
ByteArray type, which is a type from the core library that we can use to access
individual byte indexes in an array of packed bytes31 values. This type is
defined as:

```rust
struct ByteArray {
    // Full "words" of 31 bytes each. The first byte of each word in the byte array
    // is the most significant byte in the word.
    data: Array<bytes31>,
    // This felt252 actually represents a bytes31, with < 31 bytes.
    // It is represented as a felt252 to improve performance of building the byte array.
    // The number of bytes in here is specified in `pending_word_len`.

>     // The first byte is the most significant byte among the `pending_word_len` bytes in the word.

Author: This is a direct copypasta of the corelib code. I think what it means is that for example in the value 0xABCDEF, the first byte is the MSB, so 0xAB

    pending_word: felt252,

>     // Should be in range [0, 30].

Reviewer: should it be a u8?

Author: This is how it is in the corelib

# Bytecode Storage Methods for Kakarot on Starknet

The bytecode is the compiled version of your contract, and it is what the
Kakarot EVM will execute when you call the contract. As Kakarot's state is
embedded into the Starknet chain it is deployed on, you do not really "deploy"
an EVM contract on Kakarot: what actually happens is that the EVM bytecode of

> the deployed contract if first ran, and the returned data is then stored

Reviewer: if -> is 
typo

> on-chain at a particular storage address in the KakarotCore contract storage.

Reviewer: i think this flow is well explained, but might benefit from a mermaid diagram for clarity

> The Kakarot EVM will be able to load this bytecode by querying it's own storage

Reviewer: it's -> its
